File: solutions/0210-course-schedule-ii/solution.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = [[] for _ in range(numCourses)]
        indegree = [0] * numCourses

        for p in prerequisites:
            adj[p[1]].append(p[0])
            indegree[p[0]] += 1

        q = deque()
        for i in range(numCourses):
            if indegree[i] == 0:
                q.append(i)

        count = 0
        order = []
        while q:
            curr = q.popleft()
            count += 1
            order.append(curr)
            for next in adj[curr]:
                indegree[next] -= 1
                if indegree[next] == 0:
                    q.append(next)

        if count < numCourses:
            return []

        return order

File: solutions/0210-course-schedule-ii/solution.java
import java.util.*;

class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adj = new ArrayList<>();
        int[] indegree = new int[numCourses];

        for (int i = 0; i < numCourses; i++) {
            adj.add(new ArrayList<>());
        }

        for (int[] p : prerequisites) {
            adj.get(p[1]).add(p[0]);
            indegree[p[0]]++;
        }

        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                q.offer(i);
            }
        }

        int count = 0;
        int[] order = new int[numCourses];

        while (!q.isEmpty()) {
            int curr = q.poll();
            order[count] = curr;
            count++;

            for (int next : adj.get(curr)) {
                indegree[next]--;
                if (indegree[next] == 0) {
                    q.offer(next);
                }
            }
        }

        if (count < numCourses) {
            return new int[0];
        }

        return order;
    }
}

